components/style/selector_matching.rs

```diff
@@ -574,3 +583,318 @@ impl PerPseudoElementSelectorMap {
         }
     }
 }
+
+/// Map element data to Rules whose last simple selector starts with them.
+///
+/// e.g.,
+/// "p > img" would go into the set of Rules corresponding to the
+/// element "img"
+/// "a .foo .bar.baz" would go into the set of Rules corresponding to
+/// the class "bar"
+///
+/// Because we match Rules right-to-left (i.e., moving up the tree
+/// from an element), we need to compare the last simple selector in the
+/// Rule with the element.
+///
+/// So, if an element has ID "id1" and classes "foo" and "bar", then all
+/// the rules it matches will have their last simple selector starting
+/// either with "#id1" or with ".foo" or with ".bar".
+///
+/// Hence, the union of the rules keyed on each of element's classes, ID,
+/// element name, etc. will contain the Rules that actually match that
+/// element.
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub struct SelectorMap {
+    // TODO: Tune the initial capacity of the HashMap
+    id_hash: HashMap<Atom, Vec<Rule>>,
```

Comment:
The rust-selectors hashmap used fnv if I'm not wrong. This would switch us to use sip-hash, and I'd expec this to be a severe perf impact. Can you switch it back to fnv?